python/simulation.py

```diff
@@ -859,6 +864,155 @@ def amplitude(self, point, component):
 
 
+class Pade:
+    """
+    Pade is implemented as a class with a [`__call__`](#Pade.__call__) method,
+    which allows it to be used as a step function that collects field data from a given
+    point and runs [Pade](https://docs.scipy.org/doc/scipy/reference/generated/scipy.interpolate.pade.html)
+    on that data to extract an analytic rational function which models the frequency response.
+
+    See [`__init__`](#Pade.__init__) for details about constructing a `Pade`.
+
+    In particular, Pade stores the discrete time series $\\hat{f}[n]$ corresponding to the given field
+    component as a function of time and expresses it as:
+
+    $$\\hat{f}(\\omega) = \\sum_n \\hat{f}[n] e^{i\\omega n \\Delta t}$$
+
+    The above is a "Taylor-like" polynomial in $$n$$ with a Fourier basis and
+    coefficients which are the sampled field data. We then compute the Pade approximant
+    to the analytic form of this function as:
+
+    $$R(\\omega) = \\frac{P(\\omega)}{Q(\\omega)}$$
+
+    Where $$P$$ and $$Q$$ are polynomials of degree $$m$$ and $$n$$, and $$m + n + 1$$ is the
+    degree of agreement of the Pade Approximant to the analytic function $$f(\\omega)$$. This
+    function $$R$$ is stored in the callable method `pade_instance.freq_response`.
+    Be sure to save a reference to the `Pade` instance if you wish
+    to use the results after the simulation:
+
+    ```py
+    sim = mp.Simulation(...)
+    p = mp.Pade(...)
+    sim.run(p, until=time)
+    # do something with p.freq_response
+    ```
+    """
+
+    def __init__(
+        self,
+        c: int = None,
+        pt: Vector3Type = None,
+        m: Optional[float] = None,
+        n: Optional[float] = None,
+        m_frac: float = 0.5,
+        n_frac: Optional[float] = None,
+        sample_rate: int = 1,
+        start_time: Optional[int] = 0,
+        stop_time: Optional[int] = None,
+    ):
+        """
+        Construct a Pade object.
+
+        A `Pade` is a step function that collects data from the field component `c`
+        (e.g. $E_x$, etc.) at the given point `pt` (a `Vector3`). Then, at the end
+        of the run, it uses the scipy pade algorithm to approximate the analytic
+        frequency response at the specified point.
+
+        + **`c` [`Component`]** — Specifies the field component to use for extrapolation.
+         No default.
+        + **`pt` [`Vector3`]** — Specifies the location to accumulate fields. No default.
+        + **`m` [`Optional[float]`]** — Specifies the order of the numerator $$P$$. Defaults
+         to half the length of the sampled field data.
+        + **`n` [`Optional[float]`]** — Specifies the order of the denominator $$Q$$. Defaults
+         to length of sample data - m - 2.
+        + **`m_frac` [`float`]** — Alternative method for specifying `m` as a fraction of
+         field samples to use as order for numerator. Default 0.5.
+        + **`n_frac` [`Optional[float]`]** — Fraction of field samples to use as order for
```

Comment:
Suggestion: rather than have separate keyword arguments (`m`, `n`) and (`m_frac`, `n_frac`) which describe the same thing, it might improve the API to have just (`m`, `n`) and then within the constructor use the *type* of `m` and `n`  (either `int` `float` type) to infer how to define these two values.

Reply:
Great suggestion, very elegant solution. This is updated in https://github.com/NanoComp/meep/pull/2440/commits/711c4500d1329368e9075d2b2aa16e08d7000320